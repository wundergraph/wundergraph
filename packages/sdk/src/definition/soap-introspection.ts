import { ApiIntrospectionOptions, GraphQLApi } from './index';
import { introspectWithCache } from './introspection-cache';
import { IHeadersBuilder } from './headers-builder';
import { readFile } from './openapi-introspection';
import { SOAPLoader } from '@omnigraph/soap';
import { fetch } from '@whatwg-node/fetch';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
import { introspectGraphql } from './graphql-introspection';
import { WgEnv } from '../configure/options';

export interface SoapIntrospectionFile {
	kind: 'file';
	filePath: string;
}

export type SoapIntrospectionSource = SoapIntrospectionFile;

export interface SoapIntrospection {
	id?: string;
	apiNamespace?: string;
	headers?: (builder: IHeadersBuilder) => IHeadersBuilder;
	source: SoapIntrospectionSource;
}

export const introspectSoap = async (introspection: SoapIntrospection) => {
	const spec = await readFile(introspection.source.filePath);
	const configuration = { keyInput: spec, source: 'localFilesystem' };
	return introspectWithCache(
		introspection,
		configuration,
		async (introspection: SoapIntrospection, options: ApiIntrospectionOptions): Promise<GraphQLApi> => {
			return await soapToGraphQLApi(spec, introspection, options.apiID!);
		}
	);
};

const soapToGraphQLApi = async (wsdl: string, introspection: SoapIntrospection, apiID: string): Promise<GraphQLApi> => {
	const soapLoader = new SOAPLoader({
		fetch,
	});

	await soapLoader.loadWSDL(wsdl);
	const graphQLSchema = soapLoader.buildSchema();

	const schema = printSchemaWithDirectives(graphQLSchema);

	return introspectGraphql(
		{
			url: `${WgEnv.ServerUrl}-soap`,
			path: `/soap/${apiID}`,
			apiNamespace: introspection.apiNamespace,
			internal: true,
			loadSchemaFromString: () => schema,
			headers: introspection.headers,
			customIntScalars: ['BigInt'],
		},
		{}
	);
};
