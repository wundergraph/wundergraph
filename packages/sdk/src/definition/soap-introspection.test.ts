import path from 'path';
import { introspectSoap } from './soap-introspection';

describe('introspectSoap', () => {
	describe('introspection kind:file', function () {
		const fixturePath = path.join(__dirname, 'testdata', 'soap');

		const files = ['greeting.wsdl'];

		const runTest = (filePath: string) => {
			return introspectSoap({
				source: {
					kind: 'file',
					filePath,
				},
			});
		};

		files.forEach((file) => {
			it(`should introspect wsdl ${file} succesfully`, async () => {
				const filePath = path.join(fixturePath, file);
				const result = await (await runTest(filePath))({});
				expect(result.DataSources[0].Custom.UpstreamSchema).toMatchSnapshot();
				expect(result.Schema).toMatchSnapshot('wg_schema');
			});
		});
	});
});
