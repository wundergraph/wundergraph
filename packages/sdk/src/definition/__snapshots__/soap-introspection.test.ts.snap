// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`introspectSoap introspection kind:file should introspect wsdl greeting.wsdl succesfully 1`] = `
"schema {
  query: Query
  mutation: Mutation
}

directive @soap(bindingNamespace: String, elementName: String, endpoint: String) on FIELD_DEFINITION

"The \`BigInt\` scalar type represents non-fractional signed whole numeric values."
scalar BigInt

type Mutation {
  NumberConversion_NumberConversion_NumberConversionSoap_NumberToDollars(NumberToDollars: NumberConversion_NumberToDollars_Input): NumberConversion_NumberToDollarsResponse @soap(elementName: "NumberToDollarsResponse", bindingNamespace: "http://www.dataaccess.com/webservicesserver/", endpoint: "https://www.dataaccess.com/webservicesserver/NumberConversion.wso")
  NumberConversion_NumberConversion_NumberConversionSoap_NumberToWords(NumberToWords: NumberConversion_NumberToWords_Input): NumberConversion_NumberToWordsResponse @soap(elementName: "NumberToWordsResponse", bindingNamespace: "http://www.dataaccess.com/webservicesserver/", endpoint: "https://www.dataaccess.com/webservicesserver/NumberConversion.wso")
}

type NumberConversion_NumberToDollarsResponse {
  NumberToDollarsResult: String!
}

input NumberConversion_NumberToDollars_Input {
  dNum: Float
}

type NumberConversion_NumberToWordsResponse {
  NumberToWordsResult: String!
}

input NumberConversion_NumberToWords_Input {
  ubiNum: BigInt
}

type Query {
  placeholder: Void
}

"Represents NULL values"
scalar Void"
`;

exports[`introspectSoap introspection kind:file should introspect wsdl greeting.wsdl succesfully: wg_schema 1`] = `
"schema {
  query: Query
  mutation: Mutation
}

type Mutation {
  NumberConversion_NumberConversion_NumberConversionSoap_NumberToDollars(NumberToDollars: NumberConversion_NumberToDollars_Input): NumberConversion_NumberToDollarsResponse
  NumberConversion_NumberConversion_NumberConversionSoap_NumberToWords(NumberToWords: NumberConversion_NumberToWords_Input): NumberConversion_NumberToWordsResponse
}

type NumberConversion_NumberToDollarsResponse {
  NumberToDollarsResult: String!
}

input NumberConversion_NumberToDollars_Input {
  dNum: Float
}

type NumberConversion_NumberToWordsResponse {
  NumberToWordsResult: String!
}

input NumberConversion_NumberToWords_Input {
  ubiNum: Int
}

type Query {
  placeholder: Void
}

"Represents NULL values"
scalar Void"
`;
